classes/search/cmsfield.php

```diff
@@ -0,0 +1,231 @@
+     * @return \moodle_recordset|null Recordset (or null if no results)
+     */
+    public function get_document_recordset($modifiedfrom = 0, \context $context = null) {
+        global $DB;
+
+        list ($contextjoin, $contextparams) = $this->get_context_restriction_sql(
+                $context, 'cms', 'mc');
+        if ($contextjoin === null) {
+            return null;
+        }
+
+        // Search area is from customfield_data, but if the record is missing from activity, use default value.
+        $sql = "SELECT mc.id, mc.course AS courseid, mc.typeid, mcf.name AS fieldname, mcf.type,
+                       mcd.id dataid, mcd.value AS value, mcd.valueformat AS valueformat,
+                       mcd.timecreated AS timecreated, mcd.timemodified AS timemodified
+                  FROM {cms} mc
+                  JOIN {customfield_data} mcd ON mc.id = mcd.instanceid
+                  JOIN {customfield_field} mcf ON mcf.id = mcd.fieldid
+                  JOIN {customfield_category} mcc ON mcf.categoryid = mcc.id
+          $contextjoin
+                 WHERE mcd.timemodified >= ? AND mcc.component = 'mod_cms' AND mcc.area = 'cmsfield'
+                   AND mcf.type IN ('textarea', 'text')
+                 UNION
+                SELECT mc.id, mc.course AS courseid, mc.typeid, null AS fieldname, null dataid,
+                       null AS value, null AS valueformat,
+                       mc.timecreated timecreated, mc.timemodified timemodified
+                 FROM {cms} mc
+            LEFT JOIN {customfield_data} mcd ON mc.id = mcd.instanceid
+                WHERE mcd.id IS NULL AND mc.timecreated >= ?
+             ORDER BY timemodified ASC";
+        return $DB->get_recordset_sql($sql, array_merge($contextparams, [$modifiedfrom, $modifiedfrom]));
+    }
+
+    /**
+     * Returns the document associated with this data id.
+     *
+     * @param stdClass $record
+     * @param array    $options
+     * @return \core_search\document
+     */
+    public function get_document($record, $options = []) {
+        global $DB;
+        try {
+            $cm = $this->get_cm('cms', $record->id, $record->courseid);
+            $context = \context_module::instance($cm->id);
+        } catch (\dml_missing_record_exception $ex) {
+            // Notify it as we run here as admin, we should see everything.
+            debugging('Error retrieving ' . $this->areaid . ' ' . $record->id . ' document, not all required data is available: ' .
+                $ex->getMessage(), DEBUG_DEVELOPER);
+            return false;
+        } catch (\dml_exception $ex) {
+            // Notify it as we run here as admin, we should see everything.
+            debugging('Error retrieving ' . $this->areaid . ' ' . $record->id . ' document: ' . $ex->getMessage(), DEBUG_DEVELOPER);
+            return false;
+        }
+
+        // Get default value for cms custom field.
+        if (is_null($this->defaultvalues)) {
```

Comment:
If you abstract this part to get_default_values method it would be much cleaner.

Reply:
This part has been moved to the function `get_default_values`